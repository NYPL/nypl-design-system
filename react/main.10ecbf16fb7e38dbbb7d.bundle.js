(window.webpackJsonp=window.webpackJsonp||[]).push([[0],{1292:function(n,o,c){"use strict";c.r(o);c(1293),c(1323),c(1335),c(1442),c(1485)},1448:function(n,o){},1449:function(n,o){},1450:function(n,o){},1451:function(n,o){},501:function(n,o,c){c(502),c(645),n.exports=c(1292)},564:function(n,o){}},[[501,1,2]]]);