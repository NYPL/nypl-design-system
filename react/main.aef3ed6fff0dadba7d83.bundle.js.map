{"version":3,"file":"main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":"AAAA","sourceRoot":""}