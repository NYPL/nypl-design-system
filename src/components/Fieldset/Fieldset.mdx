import { ArgTypes, Canvas, Description, Meta } from "@storybook/blocks";
import { changelogData } from "./fieldsetChangelogData";
import ComponentChangelogTable from "../../utils/ComponentChangelogTable";

import * as FieldsetStories from "./Fieldset.stories";
import { Link } from "../Link/Link";

<Meta of={FieldsetStories} />

# Fieldset

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.25.3`   |
| Latest            | `3.3.2`    |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#changelog" target="_self">Changelog</Link>}

## Overview

<Description of={FieldsetStories} />

## Component Props

<Canvas of={FieldsetStories.WithControls} />

<ArgTypes of={FieldsetStories.WithControls} />

## Accessibility

This is a simple component but is very important to use when building web forms.
Syntactically, the `Fieldset` component is used to group form elements. Semantically,
the form elements that are group should have a similar theme or context. For
example, when building out a form for a user to create an account, one `Fieldset`
should be used to group the user's first and last names and another `Fieldset`
to group the user's address information.

The Reservoir DS implements the `Fieldset` component in the following form
components:

- `CheckboxGroup`, `DatePicker`, `RadioGroup`.

Resources:

- [W3C Grouping form controls with FIELDSET and LEGEND](https://www.w3.org/TR/2008/WD-WCAG20-TECHS-20080430/H82.html)
- [GOV.UK Using the fieldset and legend elements](https://accessibility.blog.gov.uk/2016/07/22/using-the-fieldset-and-legend-elements/)
- [MDN fieldset: The Field Set element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset)

## Changelog

<ComponentChangelogTable changelogData={changelogData} />
