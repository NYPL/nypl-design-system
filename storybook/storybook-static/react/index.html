<!doctype html><html lang="en"><head><meta charset="utf-8"><title>Storybook</title><meta name="viewport" content="width=device-width,initial-scale=1"><style>html, body {
    overflow: hidden;
    height: 100%;
    width: 100%;
    margin: 0;
    padding: 0;
  }</style><script>/* globals window */
  /* eslint-disable no-underscore-dangle */
  try {
    if (window.top !== window) {
      window.__REACT_DEVTOOLS_GLOBAL_HOOK__ = window.top.__REACT_DEVTOOLS_GLOBAL_HOOK__;
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('unable to connect to top frame for connecting dev tools');
  }</script><style>#root[hidden],
      #docs-root[hidden] {
        display: none !important;
      }</style></head><body><div id="root"></div><div id="docs-root"></div><script>window['DOCS_MODE'] = false;</script><script src="runtime~main.00bd9d033d7a0b972fc9.bundle.js"></script><script src="vendors~main.d2b6aafef1a259d165c0.bundle.js"></script><script src="main.3bf71672a6dd35499004.bundle.js"></script></body></html>