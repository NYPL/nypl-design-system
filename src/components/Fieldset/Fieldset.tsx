import React from "react";
import { Box, useMultiStyleConfig } from "@chakra-ui/react";

interface FieldsetProps {
  /** Children to render. Typically form-related components are used. */
  children: React.ReactNode;
  /** Additional class name to add. */
  className?: string;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** Flag to show or hide the text in the `legend` element. False by default. */
  isLegendHidden?: boolean;
  /** Flag to render "Required" or "Optional" in the `legend`.
   * False/"Optional" by default. */
  isRequired?: boolean;
  /** Text to display in the `legend` element. */
  legendText?: string;
  /** Flag to show or hide the "Required"/"Optional" text in the `legend`.
   * True by default. */
  optReqFlag?: boolean;
}

/**
 * A wrapper component that renders a `fieldset` element along with a `legend`
 * element as its first child. Commonly used to wrap form components.
 */
const Fieldset: React.FC<FieldsetProps> = ({
  children,
  className,
  id,
  isLegendHidden = false,
  isRequired = false,
  legendText,
  optReqFlag = true,
}) => {
  const styles = useMultiStyleConfig("Fieldset", { isLegendHidden });
  return (
    <Box as="fieldset" id={id} __css={styles} className={className}>
      <legend>
        {legendText}
        {optReqFlag && (
          <Box __css={styles.helper}>
            {isRequired ? "Required" : "Optional"}
          </Box>
        )}
      </legend>
      {children}
    </Box>
  );
};

export default Fieldset;
