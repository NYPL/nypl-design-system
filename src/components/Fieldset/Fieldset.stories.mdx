import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Description,
} from "@storybook/addon-docs/blocks";
import { withDesign } from "storybook-addon-designs";

import Fieldset from "./Fieldset";
import { getCategory } from "../../utils/componentCategories";
import DSProvider from "../../theme/provider";

<Meta
  title={getCategory("Fieldset")}
  component={Fieldset}
  decorators={[withDesign]}
  parameters={{
    jest: ["Fieldset.test.tsx"],
  }}
  argTypes={{
    children: { table: { disable: true } },
    className: { table: { disable: true } },
    id: { table: { disable: true } },
  }}
/>

# Fieldset

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.25.2`   |
| Latest            | `0.25.2`   |

<Description of={Fieldset} />

<Canvas withToolbar>
  <Story
    name="Basic"
    args={{
      isLegendHidden: false,
      isRequired: false,
      legendText: "This is the legend text",
      optReqFlag: true,
    }}
  >
    {(args) => (
      <Fieldset {...args}>
        <p>Text children in the Fieldset component.</p>
      </Fieldset>
    )}
  </Story>
</Canvas>

<ArgsTable story="Basic" />
