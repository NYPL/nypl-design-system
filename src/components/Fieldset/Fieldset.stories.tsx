import type { Meta, StoryObj } from "@storybook/react";
import { withDesign } from "storybook-addon-designs";

import Fieldset from "./Fieldset";

const FieldsetTemplate: Story = {
  render: ({ ...args }) => {
    return (
      <Fieldset {...args}>
        <p>Text children in the Fieldset component.</p>
      </Fieldset>
    );
  },
};

const meta: Meta<typeof Fieldset> = {
  title: "Components/Form Elements/Fieldset",
  component: Fieldset,
  decorators: [withDesign],
  argTypes: {
    id: { control: false },
    isLegendHidden: { table: { defaultValue: { summary: false } } },
    isRequired: { table: { defaultValue: { summary: false } } },
    legendText: { control: { type: "text" } },
    showRequiredLabel: { table: { defaultValue: { summary: false } } },
  },
};

export default meta;
type Story = StoryObj<typeof Fieldset>;

/**
 * Main Story for the Fieldset component. This must contains the `args`
 * and `parameters` properties in this object.
 */

export const WithControls: Story = {
  ...FieldsetTemplate,
  args: {
    id: "fieldset-id",
    isLegendHidden: false,
    isRequired: false,
    legendText: "This is the legend text",
    showRequiredLabel: false,
  },
  parameters: {
    jest: "Fieldset.test.tsx",
  },
};
