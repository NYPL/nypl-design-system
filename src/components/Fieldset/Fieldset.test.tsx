import * as React from "react";
import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import renderer from "react-test-renderer";

import Fieldset from "./Fieldset";

describe("Fieldset Accessibility", () => {
  it("passes axe accessibility", async () => {
    const { container } = render(
      <Fieldset legendText="Legend Text">
        <p>Some other fields</p>
      </Fieldset>
    );
    expect(await axe(container)).toHaveNoViolations();
  });

  it("passes axe accessibility with the legend hidden", async () => {
    const { container } = render(
      <Fieldset legendText="Legend Text" isLegendHidden>
        <p>Some other fields</p>
      </Fieldset>
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("Fieldset", () => {
  it("renders text in a legend and fieldset along with its children", () => {
    render(
      <Fieldset legendText="Legend Text">
        <p>Some other fields</p>
      </Fieldset>
    );

    expect(screen.getByRole("group")).toBeInTheDocument();
    // `legend` has no HTML role
    expect(screen.getByText(/Legend Text/i)).toBeInTheDocument();
    expect(screen.getByText("Some other fields")).toBeInTheDocument();
  });

  it("renders the UI snapshot correctly", () => {
    const basic = renderer
      .create(
        <Fieldset legendText="Legend Text">
          <p>Some other fields</p>
        </Fieldset>
      )
      .toJSON();
    const required = renderer
      .create(
        <Fieldset legendText="Legend Text" isRequired>
          <p>Some other fields</p>
        </Fieldset>
      )
      .toJSON();
    const hiddenHelperText = renderer
      .create(
        <Fieldset legendText="Legend Text" isRequired optReqFlag={false}>
          <p>Some other fields</p>
        </Fieldset>
      )
      .toJSON();
    const hiddenLegend = renderer
      .create(
        <Fieldset legendText="Legend Text" isLegendHidden>
          <p>Some other fields</p>
        </Fieldset>
      )
      .toJSON();

    expect(basic).toMatchSnapshot();
    expect(required).toMatchSnapshot();
    expect(hiddenHelperText).toMatchSnapshot();
    expect(hiddenLegend).toMatchSnapshot();
  });
});
