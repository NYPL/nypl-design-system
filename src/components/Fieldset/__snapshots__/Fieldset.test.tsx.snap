// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fieldset renders the UI snapshot correctly 1`] = `
<fieldset
  className="css-0"
  id="fieldset"
>
  <legend>
    Legend Text
  </legend>
  <p>
    Some other fields
  </p>
</fieldset>
`;

exports[`Fieldset renders the UI snapshot correctly 2`] = `
<fieldset
  className="css-0"
  id="fieldset"
>
  <legend>
    Legend Text
    <span>
       (Required)
    </span>
  </legend>
  <p>
    Some other fields
  </p>
</fieldset>
`;

exports[`Fieldset renders the UI snapshot correctly 3`] = `
<fieldset
  className="css-0"
  id="fieldset"
>
  <legend>
    Legend Text
  </legend>
  <p>
    Some other fields
  </p>
</fieldset>
`;

exports[`Fieldset renders the UI snapshot correctly 4`] = `
<fieldset
  className="css-0"
  id="fieldset"
>
  <legend>
    Legend Text
  </legend>
  <p>
    Some other fields
  </p>
</fieldset>
`;
